File: git/sqlite-crud/crud.java
import java.sql.*;

/*
https://github.com/xerial/sqlite-jdbc/releases
java -classpath ".:sqlite-jdbc-3.42.0.0.jar" Sample.java
*/

public class CRUD {
    public static void main(String[] args) {
        try {
            Connection conn = DriverManager.getConnection("jdbc:sqlite:data.db");
            System.out.println("Opened database successfully");

            Statement stmt = conn.createStatement();
            String createTableQuery = "CREATE TABLE IF NOT EXISTS Users " +
                    "(ID INT PRIMARY KEY NOT NULL, " +
                    "NAME TEXT NOT NULL, " +
                    "AGE INT NOT NULL, " +
                    "GENDER TEXT, " +
                    "SALARY INT)";
            stmt.executeUpdate(createTableQuery);
            System.out.println("Table created successfully");

            while (true) {
                System.out.println("1). Insert Records");
                System.out.println("2). Read Records");
                System.out.println("3). Update Records");
                System.out.println("4). Delete Records");
                System.out.println("5). Exit");
                int choice = Integer.parseInt(System.console().readLine("Enter Your Choice: "));

                if (choice == 1) {
                    insert(conn);
                } else if (choice == 2) {
                    System.out.println("1). Read Single Record");
                    System.out.println("2). Read All Records");
                    int readChoice = Integer.parseInt(System.console().readLine("Enter Your Choice: "));
                    if (readChoice == 1) {
                        readOne(conn);
                    } else if (readChoice == 2) {
                        readAll(conn);
                    } else {
                        System.out.println("Wrong Choice Entered");
                    }
                } else if (choice == 3) {
                    update(conn);
                } else if (choice == 4) {
                    delete(conn);
                } else if (choice == 5) {
                    break;
                } else {
                    System.out.println("Enter Correct Choice");
                }
            }

            conn.close();
        } catch (Exception e) {
            System.err.println(e.getClass().getName() + ": " + e.getMessage());
        }
    }

    public static void insert(Connection conn) {
        try {
            PreparedStatement pstmt = conn.prepareStatement("INSERT INTO Users (ID, NAME, AGE, GENDER, SALARY) VALUES (?, ?, ?, ?, ?)");

            int id = Integer.parseInt(System.console().readLine("Enter ID: "));
            String name = System.console().readLine("Enter Name: ");
            int age = Integer.parseInt(System.console().readLine("Enter Age: "));
            String gender = System.console().readLine("Enter Gender: ");
            int salary = Integer.parseInt(System.console().readLine("Enter Salary: "));

            pstmt.setInt(1, id);
            pstmt.setString(2, name);
            pstmt.setInt(3, age);
            pstmt.setString(4, gender);
            pstmt.setInt(5, salary);

            int rowsAffected = pstmt.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("Data Inserted Successfully");
            } else {
                System.out.println("Data not Inserted");
            }

            pstmt.close();
        } catch (Exception e) {
            System.err.println(e.getClass().getName() + ": " + e.getMessage());
        }
    }

    public static void readOne(Connection conn) {
        try {
            PreparedStatement pstmt = conn.prepareStatement("SELECT * FROM Users WHERE ID = ?");
            int id = Integer.parseInt(System.console().readLine("Enter Your ID: "));
            pstmt.setInt(1, id);

            ResultSet rs = pstmt.executeQuery();
            if (rs.next()) {
                System.out.println("Name is: " + rs.getString("NAME"));
                System.out.println("Age is: " + rs.getInt("AGE"));
                System.out.println("Salary is: " + rs.getInt("SALARY"));
            } else {
                System.out.println("ID Does not Exist");
            }

            rs.close();
            pstmt.close();
        } catch (Exception e) {
            System.err.println(e.getClass().getName() + ": " + e.getMessage());
        }
    }

    public static void readAll(Connection conn) {
        try {
            Statement stmt = conn.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM Users");

            System.out.println("\n<=== Available Records ===>");
            while (rs.next()) {
                System.out.println("Name is: " + rs.getString("NAME"));
                System.out.println("Age is: " + rs.getInt("AGE"));
                System.out.println("Salary is: " + rs.getInt("SALARY") + "\n");
            }

            rs.close();
            stmt.close();
        } catch (Exception e) {
            System.err.println(e.getClass().getName() + ": " + e.getMessage());
        }
    }

    public static void update(Connection conn) {
        try {
            PreparedStatement pstmt = conn.prepareStatement("UPDATE Users SET NAME = ?, AGE = ?, GENDER = ?, SALARY = ? WHERE ID = ?");

            int id = Integer.parseInt(System.console().readLine("Enter ID: "));
            String name = System.console().readLine("Enter Name: ");
            int age = Integer.parseInt(System.console().readLine("Enter Age: "));
            String gender = System.console().readLine("Enter Gender: ");
            int salary = Integer.parseInt(System.console().readLine("Enter Salary: "));

            pstmt.setString(1, name);
            pstmt.setInt(2, age);
            pstmt.setString(3, gender);
            pstmt.setInt(4, salary);
            pstmt.setInt(5, id);

            int rowsAffected = pstmt.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("Records Updated");
            } else {
                System.out.println("Something Error in Updation");
            }

            pstmt.close();
        } catch (Exception e) {
            System.err.println(e.getClass().getName() + ": " + e.getMessage());
        }
    }

    public static void delete(Connection conn) {
        try {
            PreparedStatement pstmt = conn.prepareStatement("DELETE FROM Users WHERE ID = ?");

            int id = Integer.parseInt(System.console().readLine("Enter ID: "));
            pstmt.setInt(1, id);

            int rowsAffected = pstmt.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("One record Deleted");
            } else {
                System.out.println("Something Error in Deletion");
            }

            pstmt.close();
        } catch (Exception e) {
            System.err.println(e.getClass().getName() + ": " + e.getMessage());
        }
    }
}

File: git/sqlite-crud/crud.c
#include <iostream>
#include <sqlite3.h>

using namespace std;

void create() {
    sqlite3 *db;
    int rc = sqlite3_open("data.db", &db);
    if (rc) {
        cout << "Can't open database: " << sqlite3_errmsg(db) << endl;
        return;
    }
    cout << "Opened database successfully" << endl;
    
    const char *sql = "CREATE TABLE Users (ID INT PRIMARY KEY NOT NULL, NAME TEXT NOT NULL, AGE INT NOT NULL, GENDER TEXT, SALARY INT);";
    rc = sqlite3_exec(db, sql, 0, 0, 0);
    if (rc != SQLITE_OK) {
        cout << "SQL error: " << sqlite3_errmsg(db) << endl;
    } else {
        cout << "Table created successfully" << endl;
    }
    
    sqlite3_close(db);
}

void insert() {
    sqlite3 *db;
    int rc = sqlite3_open("data.db", &db);
    if (rc) {
        cout << "Can't open database: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    sqlite3_stmt *stmt;
    const char *sql = "INSERT INTO USERS (ID, NAME, AGE, GENDER, SALARY) VALUES (?, ?, ?, ?, ?);";
    rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        cout << "SQL error: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    while (true) {
        int idd, age, salary;
        string name, gender;
        
        cout << "Enter ID: ";
        cin >> idd;
        cout << "Enter Name: ";
        cin >> name;
        cout << "Enter Age: ";
        cin >> age;
        cout << "Enter your Gender: ";
        cin >> gender;
        cout << "Enter your Salary: ";
        cin >> salary;
        
        sqlite3_bind_int(stmt, 1, idd);
        sqlite3_bind_text(stmt, 2, name.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_int(stmt, 3, age);
        sqlite3_bind_text(stmt, 4, gender.c_str(), -1, SQLITE_STATIC);
        sqlite3_bind_int(stmt, 5, salary);
        
        rc = sqlite3_step(stmt);
        if (rc != SQLITE_DONE) {
            cout << "Data not Inserted" << endl;
        } else {
            cout << "Data Inserted Successfully" << endl;
        }
        
        sqlite3_reset(stmt);
        
        char ch;
        cout << "Do You want to Add More Records(Y/N): ";
        cin >> ch;
        if (ch == 'N' || ch == 'n') {
            break;
        }
    }
    
    sqlite3_finalize(stmt);
    sqlite3_close(db);
}

void read_one() {
    sqlite3 *db;
    int rc = sqlite3_open("data.db", &db);
    if (rc) {
        cout << "Can't open database: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    sqlite3_stmt *stmt;
    const char *sql = "SELECT * FROM USERS WHERE id = ?;";
    rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        cout << "SQL error: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    int ids;
    cout << "Enter Your ID: ";
    cin >> ids;
    
    sqlite3_bind_int(stmt, 1, ids);
    
    rc = sqlite3_step(stmt);
    if (rc == SQLITE_ROW) {
        cout << "Name is: " << sqlite3_column_text(stmt, 1) << endl;
        cout << "Age is: " << sqlite3_column_int(stmt, 2) << endl;
        cout << "Salary is: " << sqlite3_column_int(stmt, 4) << endl;
    } else {
        cout << "Record Does not Exist" << endl;
    }
    
    sqlite3_finalize(stmt);
    sqlite3_close(db);
}

void read_all() {
    sqlite3 *db;
    int rc = sqlite3_open("data.db", &db);
    if (rc) {
        cout << "Can't open database: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    sqlite3_stmt *stmt;
    const char *sql = "SELECT * FROM USERS;";
    rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        cout << "SQL error: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    rc = sqlite3_step(stmt);
    if (rc == SQLITE_ROW) {
        cout << "\n<=== Available Records ===>" << endl;
        while (rc == SQLITE_ROW) {
            cout << "Name is: " << sqlite3_column_text(stmt, 1) << endl;
            cout << "Age is: " << sqlite3_column_int(stmt, 2) << endl;
            cout << "Salary is: " << sqlite3_column_int(stmt, 4) << endl;
            cout << endl;
            rc = sqlite3_step(stmt);
        }
    }
    
    sqlite3_finalize(stmt);
    sqlite3_close(db);
}

void update() {
    sqlite3 *db;
    int rc = sqlite3_open("data.db", &db);
    if (rc) {
        cout << "Can't open database: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    sqlite3_stmt *stmt;
    const char *sql = "UPDATE USERS SET NAME = ?, AGE = ?, GENDER = ?, SALARY = ? WHERE ID = ?;";
    rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        cout << "SQL error: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    int idd, age, salary;
    string name, gender;
    
    cout << "Enter ID: ";
    cin >> idd;
    cout << "Enter Name: ";
    cin >> name;
    cout << "Enter Age: ";
    cin >> age;
    cout << "Enter Gender: ";
    cin >> gender;
    cout << "Enter Salary: ";
    cin >> salary;
    
    sqlite3_bind_text(stmt, 1, name.c_str(), -1, SQLITE_STATIC);
    sqlite3_bind_int(stmt, 2, age);
    sqlite3_bind_text(stmt, 3, gender.c_str(), -1, SQLITE_STATIC);
    sqlite3_bind_int(stmt, 4, salary);
    sqlite3_bind_int(stmt, 5, idd);
    
    rc = sqlite3_step(stmt);
    if (rc != SQLITE_DONE) {
        cout << "Something Error in Updation" << endl;
    } else {
        cout << "Records Updated" << endl;
    }
    
    sqlite3_finalize(stmt);
    sqlite3_close(db);
}

void deleteRecord() {
    sqlite3 *db;
    int rc = sqlite3_open("data.db", &db);
    if (rc) {
        cout << "Can't open database: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    sqlite3_stmt *stmt;
    const char *sql = "DELETE FROM USERS WHERE ID = ?;";
    rc = sqlite3_prepare_v2(db, sql, -1, &stmt, 0);
    if (rc != SQLITE_OK) {
        cout << "SQL error: " << sqlite3_errmsg(db) << endl;
        return;
    }
    
    int idd;
    cout << "Enter ID: ";
    cin >> idd;
    
    sqlite3_bind_int(stmt, 1, idd);
    
    rc = sqlite3_step(stmt);
    if (rc != SQLITE_DONE) {
        cout << "Something Error in Deletion" << endl;
    } else {
        cout << "One record Deleted" << endl;
    }
    
    sqlite3_finalize(stmt);
    sqlite3_close(db);
}

int main() {
    while (true) {
        cout << "1). Insert Records" << endl;
        cout << "2). Read Records" << endl;
        cout << "3). Update Records" << endl;
        cout << "4). Delete Records" << endl;
        cout << "5). Exit" << endl;
        cout << "Enter Your Choice: ";
        
        int ch;
        cin >> ch;
        
        switch (ch) {
            case 1:
                insert();
                break;
            case 2:
                cout << "1). Read Single Record" << endl;
                cout << "2). Read All Records" << endl;
                cout << "Enter Your Choice: ";
                
                int choice;
                cin >> choice;
                
                if (choice == 1) {
                    read_one();
                } else if (choice == 2) {
                    read_all();
                } else {
                    cout << "Wrong Choice Entered" << endl;
                }
                
                break;
            case 3:
                update();
                break;
            case 4:
                deleteRecord();
                break;
            case 5:
                return 0;
            default:
                cout << "Enter Correct Choice" << endl;
                break;
        }
    }
    
    return 0;
}
